client/state/selectors/get-billing-transaction-date-filter-values.js

```diff
@@ -0,0 +1,70 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import { translate, moment } from 'i18n-calypso';
+import { forEach, orderBy, range } from 'lodash';
+/**
+ * Internal dependencies
+ */
+import createSelector from 'lib/create-selector';
+import getPastBillingTransactions from './get-past-billing-transactions';
+import getUpcomingBillingTransactions from './get-upcoming-billing-transactions';
+
+/**
+ * Based on the transactions list, returns metadata for rendering the date filters with counts
+ *
+ * @param  {Object}  state           Global state tree
+ * @param  {String}  transactionType Transaction type
+ * @return {Array}                   Date filter metadata
+ */
+export default createSelector(
+	( state, transactionType ) => {
+		const transactions =
+			'upcoming' === transactionType
+				? getUpcomingBillingTransactions( state )
+				: getPastBillingTransactions( state );
+
+		if ( ! transactions ) {
+			return [];
+		}
+
+		const result = range( 6 ).reduce( function( accumulator, n ) {
```

Comment:
Seems like the internals of this selector could be optimized a little bit - a lot is happening here and we could possibly improve it to do a little less 😉 For example, since we end up returning an array, we could start with an array; we could use `times()` instead of `range().reduce()`; we could use `tail()` of the array to get the oldest month instead of generating a new moment; reduce overall mutations we're doing; etc. 

I can see you're essentially porting this from the current transactions header, so probably it could be done in a separate PR. In case you decide to do it in another PR, maybe adding a TODO comment would make sense.